extensions/ext/GLSL_EXT_spirv_intrinsics.txt

```diff
@@ -0,0 +1,1038 @@
+        `spirv_storage_class` qualifier.
+
+        This functionality is only available when generating SPIR-V.
+
+        Any extensions or capabilities needed by the storage class must be declared
+        using the optional spirv-requirements-list as described in section 3.9
+        (SPIR-V Requirements Declarations).
+        
+        The `spirv_storage_class` qualifier can only be used on variable
+        declarations at global scope.
+
+        spirv_storage_class-qualifier:
+
+            `spirv_storage_class` ( integer-constant-expression )
+            `spirv_storage_class` ( spirv-requirements-list, integer-constant-expression )
+
+        The `spirv_storage_class` modifies the *StorageClass* operand of
+        declared variables to be the storage class defined for the
+        *StorageClass* enumeration equal to the integer value provided.
+
+        Any extensions or capabilities needed by the storage class must be
+        declared using SPIR-V Requirements Declarations (see section 3.9).
+
+    Add a new section 4.14. SPIR-V Decorations
+
+        SPIR-V decorations not known to or exposed by the GLSL compiler can
+        be added by use of the `spirv_decorate`, `spirv_decorate_id`, or
+        `spirv_decorate_string` qualifiers.
+
+        This functionality is only available when generating SPIR-V.
+
+        Any extensions or capabilities needed by the decoration must be declared
+        using the optional spirv-requirements-list as described in section 3.9
+        (SPIR-V Requirements Declarations).
+        
+        The `spirv_decorate`, `spirv_decorate_id`, and
+        `spirv_decorate_string` qualifiers can be used on variable
+        declarations, function parameters, function return types,
+        and expressions (with "constructor" syntax).
+        `spirv_decorate_id` must not be used on structure member
+        declarations, but `spirv_decorate` and `spirv_decorate_string` can.
+
+        spirv_decorate-qualifier:
+
+            `spirv_decorate` ( integer-constant, spirv_decorate-parameter-list )
+            `spirv_decorate` ( spirv-requirements-list , integer-constant, spirv_decorate-parameter-list )
+            `spirv_decorate_id` ( integer-constant, spirv_decorate_id-parameter-list  )
+            `spirv_decorate_id` ( spirv-requirements-list , integer-constant, spirv_decorate_id-parameter-list  )
+            `spirv_decorate_string` ( integer-constant, spirv_decorate_string-parameter-list )
+            `spirv_decorate_string` ( spirv-requirements-list , integer-constant, spirv_decorate_string-parameter-list )
+
+        spirv_decorate-parameter-list:
+
+            spirv_decorate-parameter
+            spirv_decorate-parameter , spirv_decorate-parameter-list
+
+        spirv_decorate-parameter:
+
+            floating-constant
+            integer-constant
+            bool-constant
+
+        spirv_decorate_id-parameter-list:
+
+            constant-expression
```

Comment:
A change is to fix a problem here (https://github.com/KhronosGroup/glslang/pull/3170). We should allow variables in the `spirv_decorate_id-parameter-list`. This is because `OpDecorateId` allows such cases.

```suggestion
    spirv_decorate_id_parameter_list
        spirv_decorate_id_parameter
        spirv_decorate_id_parameter spirv_decorate_id_parameter_list 

    spirv_decorate_id_parameter
        variable-identifier
        floating-constant
        integer-constant
        bool-constant
```

Reply:
Assuming that constant-expression in the grammar doesn't actually need constants, I think we're good. If not I might need some guidance on what grammar identifier should be in its place...

Reply:
constant-expression is an alias in the grammar for conditional-expression that gets used when things have to be constant. It's used for layout qualifiers at the moment, so if you don't intend this to follow similar rules then it should probably say conditional-expression instead. Based on @amdrexu's suggestion it sounds like primary-expression might be the most appropriate though?